### DISCUSSION TITLE
How to run a script before Renovate will lock updated dependencies (like `postUpgradeTasks`, but `preUpgradeTasks`)?

### DISCUSSION THREAD

**Discussion by simon-liebehenschel** [renovatebot/renovate#19230] [Request Help] [answered]
## My use case
I have a Renovate image 
```Dockerfile
FROM renovate/renovate:34.48-slim
SHELL ["/bin/bash", "-o", "pipefail", "-c"]
# We need to have all Python versions used by our projects.
USER root
RUN install-tool python 3.10.8  \
    && install-tool python 3.11.0 \
    && install-tool poetry 1.2.2
USER 1000
```
I have dozens of repositories 
- repository A uses Python 3.10.8
- repository B uses Python 3.11.0
- repository B uses Python ...
- ...

Before Renovate will call `poetry lock` I must call `poetry env use ${PYTHON_VERSION}` to select a Python version that is used by a current repository (e.g. _3.10.8_, or _3.11.0_).

## Question

How can I run a script per repository before Renovate will start updating a repository? A.k.a. `preUpgradeTasks`.

## Versions
- Renovate self-hosted at GitLab SaaS

---

PS:

This is something that Renovate should do on its side before calling `poetry lock`: take a look at the "pyproject.toml", parse Python version, and call `poetry env use ${PYTHON_VERSION}`. But since there are already 1000+ open issues and as a Python developer I can not contribute here, I think I need to have a workaround on my side.

**Comment by rarkins** [comment, marked as answer]
https://github.com/search?q=repo%3Arenovatebot%2Frenovate+preUpgradeTasks&type=issues
